guppylang/compiler/expr_compiler.py

```diff
@@ -452,68 +463,104 @@ def visit_ResultExpr(self, node: ResultExpr) -> Wire:
+        return self.dfg[list_place]
+
+    def visit_DesugaredArrayComp(self, node: DesugaredArrayComp) -> Wire:
+        # Allocate an uninitialised array of the desired size and a counter variable
+        array_ty = get_type(node)
+        assert isinstance(array_ty, OpaqueType)
+        array_var = Variable(next(tmp_vars), array_ty, node)
+        count_var = Variable(next(tmp_vars), int_type(), node)
+        hugr_elt_ty = (
+            ht.Option(node.elt_ty.to_hugr())
+            if node.elt_ty.linear
+            else node.elt_ty.to_hugr()
+        )
+        self.dfg[array_var] = self.builder.add_op(
+            array_new_uninitialized(hugr_elt_ty, node.length)
+        )
+        self.dfg[count_var] = self.builder.load(
+            hugr.std.int.IntVal(0, width=NumericType.INT_WIDTH)
+        )
+
+        def build_update(elt: Wire) -> None:
+            """Callback for putting an element into the array."""
+            array, count = self.dfg[array_var], self.dfg[count_var]
+            [], [self.dfg[array_var]] = self._build_method_call(
+                array_ty, "__setitem__", node, [array, count, elt], array_ty.args
+            )
+            # Update `count += 1`
+            one = self.builder.load(hugr.std.int.IntVal(1, width=NumericType.INT_WIDTH))
+            [self.dfg[count_var]], [] = self._build_method_call(
+                int_type(), "__add__", node, [count, one]
+            )
+
+        self._compile_generators(
+            node.elt, [node.generator], [array_var, count_var], build_update
+        )
+        return self.dfg[array_var]
+
+    def _build_method_call(
+        self,
+        ty: Type,
+        method: str,
+        node: AstNode,
+        args: list[Wire],
+        type_args: Sequence[Argument] | None = None,
+    ) -> CallReturnWires:
+        func = self.globals.get_instance_func(ty, method)
+        assert func is not None
+        return func.compile_call(args, type_args or [], self.dfg, self.globals, node)
+
+    def _compile_generators(
+        self,
+        elt: ast.expr,
+        gens: list[DesugaredGenerator],
+        loop_vars: list[Variable],
+        build_update: Callable[[Wire], None],
+    ) -> None:
+        """Helper function to generate nested TailLoop nodes for generators"""
+        from guppylang.compiler.stmt_compiler import StmtCompiler
+
+        compiler = StmtCompiler(self.globals)
+        with ExitStack() as stack:
+            for gen in gens:
+                # Build the generator
+                compiler.compile_stmts([gen.iter_assign], self.dfg)
+                assert isinstance(gen.iter, PlaceNode)
+                assert isinstance(gen.hasnext, PlaceNode)
+                inputs = [gen.iter] + [PlaceNode(place=var) for var in loop_vars]
+                # Remember to finalize the iterator once we are done with it. Note that
+                # we need to use partial in the callback, so that we bind the *current*
```

Comment:
This is the seriously awkward bit, which of course I hadn't spotted when I suggested it ;-), well done!